### PR TITLE
Several improvements to Elasticsearch.Ext.Logging

### PR DESCRIPTION
**Pull request by Mpdreamz**
- Add some basic integration tests to assert the logs make it in the right index
- Rename `ElasticsearchDataProcessor` to `ElasticsearchDataShipper`
- Move enrichtment of `LogEvent` out of the shipper and into static helper class.
- Rely on elasticsearch to generate ids
- Move to PostData.StreamHandler to serialize directly to the IO stream.

relates: https://github.com/elastic/ecs-dotnet/issues/95

### CODE REVIEW

File: src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+		}
+
+		public void Dispose()
+		{
+			_messageQueue?.CompleteAdding();
+			try
+			{
+				_outputThread.Join(1500); // with timeout in case writer is locked
+			}
+			catch (ThreadStateException) { }
```
Comment (russcam) [inline src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs:50]: is there an _internal_ logger to which this exception can be written?

File: src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+			{
+				// This is SingleNode with "http://localhost:9200"
+				settings = new ConnectionConfiguration();
+			}
+			else if (_options.ConnectionPoolType == ConnectionPoolType.SingleNode
+				|| _options.ConnectionPoolType == ConnectionPoolType.Unknown && nodes.Length == 1)
+				settings = new ConnectionConfiguration(nodes[0]);
+			else
+			{
+				settings = new ConnectionConfiguration(connectionPool);
```
Comment (russcam) [inline src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs:127]: Looks like `_options.ShipTo.CreateConnectionPool();` can return `null`: 

https://github.com/elastic/ecs-dotnet/blob/249b86d9163bd373182ef5d458910a98f7a7973e/src/Elasticsearch.Extensions.Logging/ShipTo.cs#L61-L86

but `ConnectionConfiguration` does not guard against `null`: 

https://github.com/elastic/elasticsearch-net/blob/f1ffbf260132eb83ea4e7950c1bcf19ef679ee6f/src/Elasticsearch.Net/Configuration/ConnectionConfiguration.cs#L210

File: src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+			}
+
+			var lowlevelClient = new ElasticLowLevelClient(settings);
+
+			_ = Interlocked.Exchange(ref _lowLevelClient, lowlevelClient);
```
Comment (russcam) [inline src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs:132]: I'm struggling to see what benefit this provides over assigning to the field directly in this case, as the original value is not used and field assignment is atomic on all .NET platforms AFAIK.  Is there a benefit that I'm missing?

File: src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+				foreach (var logEvent in _messageQueue.GetConsumingEnumerable()) PostEvent(logEvent);
+			}
+			catch
+			{
+				try
+				{
+					_messageQueue.CompleteAdding();
+				}
+				catch { }
+			}
```
Comment (russcam) [inline src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs:105]: If I understand this correctly, an exception in posting the event will mark the blocking collection as no longer accepting items, causing any subsequent items to be sent to `PostEvent` directly in `EnqueueMessage`. Would we want to distinguish the types of exceptions that should cause this behaviour vs. those that shouldn't e.g. are there exceptions the client could throw where this behaviour is undesirable?

File: src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+		private void PostEvent(LogEvent logEvent)
+		{
+			var indexTime = logEvent.Timestamp ?? ElasticsearchLoggerProvider.LocalDateTimeProvider();
+			if (_options.IndexOffset.HasValue) indexTime = indexTime.ToOffset(_options.IndexOffset.Value);
+
+			var index = string.Format(_options.Index, indexTime);
+
+			var localClient = _lowLevelClient;
+			var response = localClient.Index<StringResponse>(index,
```
Comment (russcam) [inline src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs:83]: I had a question about bulk, but see that this is addressed in #99 😄

File: tests/Elasticsearch.Extensions.Logging.IntegrationTests/LoggingTests.cs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+		[Fact]
+		public async Task LogsEndUpInCluster()
+		{
+			using var _ = CreateLogger(out var logger, out var indexPrefix);
+
+			logger.LogError("an error occured");
+
+			// TODO make sure we can await something here on ElasticsearchDataShipper
+			await Task.Delay(TimeSpan.FromSeconds(5));
```
Comment (russcam) [inline tests/Elasticsearch.Extensions.Logging.IntegrationTests/LoggingTests.cs:62]: Could the logger expose a property/event to ascertain that it has sent or is sending events?

File: src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs (hunk 1 above)
Comment (Mpdreamz) [reply src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs:50]: I looked into fallback loggers in the Mic.Ext.Logging world but there is no such concept. The best we can do is expose callbacks to allow to user to log manually to fallback locations in their application configuration.

This is a carry over of #85 (introduced there not here) that is rewritten in #98.

File: src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs (hunk 2 above)
Comment (Mpdreamz) [reply src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs:127]: Good catch!, this is a carry over from #98, will update the meta ticket #95  to follow up in a separate PR

File: src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs (hunk 4 above)
Comment (Mpdreamz) [reply src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs:105]: This is a carry over from #85 and is now rewritten to move away from `BlockingCollection` to `Channels` in #99

File: tests/Elasticsearch.Extensions.Logging.IntegrationTests/LoggingTests.cs (hunk 6 above)
Comment (Mpdreamz) [reply tests/Elasticsearch.Extensions.Logging.IntegrationTests/LoggingTests.cs:62]: That was my intend but I knew the shipper had to be rewritten so I left the TODO :) 

Once #99 lands we can expose more callbacks and stats

**Comment by Mpdreamz** [comment]
Thanks for the review @russcam I think most comments should be addressed in #99 

Noted the need for `CreateConnectionPool` not to return null in the meta ticket so we can address this in a smaller dedicated PR.

File: src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs (hunk 3 above)
Comment (Mpdreamz) [reply src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs:132]: There is not, carry over from #85 we can address that in a follow up PR.

File: src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs [hunk 7]
```diff
@@ -0,0 +1,135 @@
+// Licensed to Elasticsearch B.V under one or more agreements.
+// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information
+
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading;
+using Elastic.CommonSchema;
```
Comment (gregkalapos) [inline src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs:9]:
```suggestion
```

Comment (Mpdreamz) [reply src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs:9]: Will tackle later 👍

**Pull request merged**